src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,158 @@
+    {
+        if (this.options == null)
+        {
+            return;
+        }
+
+        var eventSourceName = eventData.EventSource.Name;
+
+        if (eventData.EventName != "EventCounters")
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreNonEventCountersName(eventSourceName);
+            return;
+        }
+
+        if (eventData.Payload == null || eventData.Payload.Count == 0 || eventData.Payload[0] is not IDictionary<string, object>)
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenEventArgsPayloadNotParseable(eventSourceName);
+            return;
+        }
+
+        var payload = eventData.Payload[0] as IDictionary<string, object>;
+        var hasName = payload.TryGetValue("Name", out var nameObj);
+
+        if (!hasName)
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenEventArgsWithoutName(eventSourceName);
+            return;
+        }
+
+        var name = nameObj.ToString();
+
+        var hasMean = payload.TryGetValue("Mean", out var mean);
+        var hasIncrement = payload.TryGetValue("Increment", out var increment);
+
+        if (!(hasIncrement ^ hasMean))
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreMeanIncrementConflict(eventSourceName);
+            return;
+        }
+
+        var value = Convert.ToDouble(hasMean ? mean : increment);
+        this.UpdateInstrumentWithEvent(hasMean, eventSourceName, name, value);
+    }
+
+    private static Dictionary<string, string> GetEnableEventsArguments(EventCountersInstrumentationOptions options) =>
+        new() { { "EventCounterIntervalSec", options.RefreshIntervalSecs.ToString() } };
+
+    private static bool IsValidInstrumentName(string name) =>
+        !string.IsNullOrWhiteSpace(name) && InstrumentNameRegex.IsMatch(name);
+
+    private void UpdateInstrumentWithEvent(bool isGauge, string eventSourceName, string name, double value)
+    {
+        try
+        {
+            ValueTuple<string, string> metricKey = new(eventSourceName, name);
+            _ = this.values.AddOrUpdate(metricKey, value, isGauge ? (_, _) => value : (_, existing) => existing + value);
+
+            var instrumentName = $"EventCounters.{eventSourceName}.{name}";
+
+            if (!IsValidInstrumentName(instrumentName))
+            {
+                EventCountersInstrumentationEventSource.Log.InvalidInstrumentNameWarning(instrumentName);
+                return;
+            }
+
+            if (!this.instruments.ContainsKey(metricKey))
+            {
+                Instrument instrument = isGauge
+                    ? MeterInstance.CreateObservableGauge(instrumentName, () => this.values[metricKey])
+                    : MeterInstance.CreateObservableCounter(instrumentName, () => this.values[metricKey]);
+                _ = this.instruments.TryAdd(metricKey, instrument);
+
+                if (!this.instrumentNames.TryAdd(instrumentName, 0))
```

Comment:
No need of checking duplicate instrument names either. These things are done centrally at the SDK setup:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MetricReaderExt.cs#L54-L61

Reply:
True, I guess I was adding these checks prematurely before I figure out a good way to map EventCounter names into instrument names.